README.md
# Pesa SDK for PHP

[![Latest Version on Packagist](https://img.shields.io/packagist/v/openpesa/pesa.svg?style=flat-square&?include_prereleases)](https://packagist.org/packages/openpesa/pesa)
[![Build Status](https://img.shields.io/travis/openpesa/php-pesa/main.svg?style=flat-square)](https://travis-ci.org/openpesa/php-pesa)
[![codecov.io](https://img.shields.io/codecov/c/github/openpesa/php-pesa/main?style=flat-square)](https://codecov.io/github/openpesa/php-pesa)
[![Total Downloads](https://img.shields.io/packagist/dt/openpesa/pesa.svg?style=flat-square)](https://packagist.org/packages/openpesa/pesa)

The **Pesa SDK for PHP** makes it easy for developers to access [OpenAPI](https://openapiportal.m-pesa.com/) in their PHP code, and build robust applications and software using services like Customber 2 Bussiness, Query etc.

## Documentation

Take a look at the [API docs here](https://php-pesa.netlify.app/).

## Getting Started

1. **Sign up for OpenAPI Portal** – Before you begin, you need to
   sign up for an account and retrieve your credentials.

1. **Minimum requirements** – To run the SDK, your system will need to meet the
   [minimum requirements](https://php-pesa.netlify.app/docs/requirements.html), including having **PHP >= 7.1**.
   <!-- We highly recommend having it compiled with the cURL extension and cURL
   7.16.2+ compiled with a TLS backend (e.g., NSS or OpenSSL). -->
1. **Install the SDK** – Using [Composer] is the recommended way to install the
   Pesa SDK for PHP. The SDK is available via [Packagist] under the
   [`openpesa/php-pesa`](https://packagist.org/packages/openpesa/pesa) package. If Composer is installed globally on your system, you can run the following in the base directory of your project to add the SDK as a dependency:
   ```sh
   composer require openpesa/pesa
   ```
   Please see the
   [Installation section of the User Guide](https://php-pesa.netlify.app/docs/installation.html) for more
   detailed information about installing the SDK through Composer and other
   means.
1. **Using the SDK** – The best way to become familiar with how to use the SDK
   is to read the [User Guide](https://php-pesa.netlify.app/docs/guide.html). 
   
   <!-- The [Getting Started Guide](#) will help you become familiar with
   the basic concepts. -->


## Usage

### Quick Examples

```php

require 'vendor/autoload.php';

use Openpesa\SDK\Forodha;

// Intiate with credentials
$forodha = new Forodha([
            'api_key' => 'YOUR_API_KEY', 
            'public_key' => 'PUBLIC_KEY', 
            'client_options' => [], 
        ]);

// Setup the transaction 
$data = [
    'input_Amount' => '10000',
    'input_Country' => 'TZN',
    'input_Currency' => 'TZS',
    'input_CustomerMSISDN' => '255766303775',
    'input_ServiceProviderCode' => '000000',
    'input_ThirdPartyConversationID' => 'rerekf',
    'input_TransactionReference' => rand(),
    'input_PurchasedItemsDesc' => 'Test Two Item'
];

// Execute 
$result = $forodha->transact('c2b', $data);

// Print results
var_dump($result);

```

For more example check [pesa-demo-example](https://github.com/openpesa/php-pesa/tree/develop/examples).

### Testing

```bash
composer test
```

## Opening Issues

If you have a feature requrest or you encounter a bug, please file an issue on [our issue tracker on GitHub](https://github.com/openpesa/php-pesa/issues).

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please review our [CONTRIBUTING](CONTRIBUTING.md) for details. 


### Security

If you discover any security related issues, please email [alphaolomi@gmail.com](mailto:alphaolomi@gmail.com) instead of using the issue tracker.

## Credits

-   [Alpha Olomi](https://github.com/openpesa)
-   [Ley](https://github.com/leyluj)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `pesa` will be documented in this file

## 0.0.1 - 2020-10-12

- initial release
- Support for C2B, B2C, REVERSAL


src/Pesa.php
<?php

namespace Openpesa\SDK;

require 'vendor/autoload.php';

use Exception;
use GuzzleHttp\Client;
use phpseclib\Crypt\RSA;

class Pesa
{
    protected $client;
    private $APIKEY = "";
    const BASE_DOMAIN         = "https://openapi.m-pesa.com";
    const BASE_SANDBOX_DOMAIN = self::BASE_DOMAIN . "/sandbox/";

    private $publicKey = "MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEArv9yxA69XQKBo24BaF/D+fvlqmGdYjqLQ5WtNBb5tquqGvAvG3WMFETVUSow/LizQalxj2ElMVrUmzu5mGGkxK08bWEXF7a1DEvtVJs6nppIlFJc2SnrU14AOrIrB28ogm58JjAl5BOQawOXD5dfSk7MaAA82pVHoIqEu0FxA8BOKU+RGTihRU+ptw1j4bsAJYiPbSX6i71gfPvwHPYamM0bfI4CmlsUUR3KvCG24rB6FNPcRBhM3jDuv8ae2kC33w9hEq8qNB55uw51vK7hyXoAa+U7IqP1y6nBdlN25gkxEA8yrsl1678cspeXr+3ciRyqoRgj9RD/ONbJhhxFvt1cLBh+qwK2eqISfBb06eRnNeC71oBokDm3zyCnkOtMDGl7IvnMfZfEPFCfg5QgJVk1msPpRvQxmEsrX9MQRyFVzgy2CWNIb7c+jPapyrNwoUbANlN8adU1m6yOuoX7F49x+OjiG2se0EJ6nafeKUXw/+hiJZvELUYgzKUtMAZVTNZfT8jjb58j8GVtuS+6TM2AutbejaCV84ZK58E2CRJqhmjQibEUO6KPdD7oTlEkFy52Y1uOOBXgYpqMzufNPmfdqqqSM4dU70PO8ogyKGiLAIxCetMjjm6FCMEA3Kc8K0Ig7/XtFm9By6VxTJK1Mg36TlHaZKP6VzVLXMtesJECAwEAAQ==";


    public function __construct()
    {
        $this->client = new Client([            
            'timeout'  => 300,
            'headers' => [
                'Accept' => 'application/json',
                'Origin' => '*'
            ]
        ]);
    }

    public function GenerateSessionKey()
    {
        $response = $this->client->request(
            'GET',
            'https://openapi.m-pesa.com/sandbox/ipg/v2/vodacomTZN/getSession/',
            ['headers' => ['Authorization' => "Bearer {$this->create_bearer_token($this->APIKEY)}"]]
        );
        $body = json_decode($response->getBody());
        // print_r($body);
        return $body->output_SessionID;
    }

    function create_bearer_token($apiKEY)
    {
        // Need to do these lines to create a 'valid' formatted RSA key for the openssl library
        $rsa = new RSA();
        $rsa->loadKey($this->publicKey);
        $rsa->setPublicKey($this->publicKey);

        $publickey = $rsa->getPublicKey();
        $api_encrypted = '';
        $encrypted = '';

        if (openssl_public_encrypt($apiKEY, $encrypted, $publickey)) {
            $api_encrypted = base64_encode($encrypted);
        }
        return $api_encrypted;
    }

    function sendUSSDPUSH()
    {   
        try {
            //code...
            $sessionID = $this->GenerateSessionKey();
        } catch (\Throwable $th) {
            //throw $th;
            print_r($th->getMessage());
        } 

        $token =   $this->create_bearer_token($sessionID);
        print_r("\n\ntoekn ". $token);
        $data = array(
            'input_Amount' => 2030,
            'input_Country' => 'TZN',
            'input_Currency' => 'TZS',
            'input_CustomerMSISDN' => '255766303775',
            'input_ServiceProviderCode' => '000000',
            'input_ThirdPartyConversationID' => 'rerekf',
            'input_TransactionReference' => 'odfdferre',
            'input_PurchasedItemsDesc' => 'Test Two Item'
        );
        try {
            //code...
            $response = $this->client->request('POST',
                'https://openapi.m-pesa.com/sandbox/ipg/v2/vodacomTZN/c2bPayment/singleStage/',
                $data,
                ['headers' => [
                    'Authorization' => "Bearer {$token}"
                ]]
            );
        } catch (\Throwable $th) {
            //throw $th;
            print_r($th->getMessage());
        }
        
        print_r(
            json_decode($response->getBody())
        );
    }
}



$pesa = new Pesa();
// $pesa->GenerateSessionKey();
$pesa->sendUSSDPUSH();
